packages/flutter_tools/lib/src/dart/package_map.dart

```diff
@@ -16,6 +16,45 @@ Future<PackageConfig> currentPackageConfig() async {
   return loadPackageConfigUri(Isolate.packageConfigSync!);
 }
 
+/// Locates the `.dart_tool/package_config.json` relevant to [dir].
+///
+/// Searches [dir] and all parent directories.
+///
+/// Returns `null` if no package_config.json was found.
+File? findPackageConfigFile(Directory dir) {
+  final FileSystem fileSystem = dir.fileSystem;
```

Comment:
nit, this can be a little cleaner by using more of the built-in APIs from package:file:

```dart
File? findPackageConfigFile(Directory dir) {
  final FileSystem fileSystem = dir.fileSystem;

  Directory candidateDir = fileSystem.directory(dir.path).absolute;
  while (true) {
    final File candidatePackageConfigFile = candidateDir
        .childDirectory('.dart_tool')
        .childFile('package_config.json');
    if (candidatePackageConfigFile.existsSync()) {
      return candidatePackageConfigFile;
    }
    // TODO(sigurdm): we should not need to check this file, it is obsolete,
    // https://github.com/flutter/flutter/issues/150908.
    final File candidatePackagesFile = candidateDir.childFile('.packages');
    if (candidatePackagesFile.existsSync()) {
      return candidatePackagesFile;
    }
    final Directory parentDir = candidateDir.parent;
    if (fileSystem.identicalSync(parentDir.path, candidateDir.path)) {
      return null;
    }
    candidateDir = parentDir;
  }
}
```

Reply:
Nice!